rfd/0127-encrypted-session-recordings.md

```diff
@@ -0,0 +1,380 @@
+  // RotatedKeys is a list of wrapped private keys that have been rotated.
+  // These are kept to decrypt historical encrypted session recordings.
+  repeated WrappedKey RotatedKeys = 3 [
+    (gogoproto.jsontag) = "rotated_keys"
+  ];
+}
+
+// SessionRecordingConfigSpecV2 is the actual data we care about
+// for SessionRecordingConfig.
+message SessionRecordingConfigSpecV2 {
+  // existing fields omitted 
+
+  BoolValue encrypted = 3 [(gogoproto.jsontag) = "encrypted"]
+}
+
+// SessionRecordingConfigV2 contains session recording configuration.
+message SessionRecordingConfigV2 {
+  // existing fields omitted
+
+  // Status contains all of the current and rotated keys used for encrypted
+  // session recording
+  SessionRecordingConfigStatusV2 Status = 6 [
+    (gogoproto.nullable) = true,
+    (gogoproto.jsontag) = "status"
+  ]
+}
+
+// api/proto/teleport/trust/v1/trust_service.proto
+
+// TrustService provides methods to manage certificate authorities.
+service TrustService {
+  // existing RPCs omitted
+
+  // RotateSessionRecordingKeys rotates the keys associated with encrypting and
+  // decrypting session recordings.
+  rpc RotateSessionRecordingKeys(RotateSessionRecordingKeysRequest) returns (RotateSessionRecordingKeysResponse);
+}
+
+// RecordingKeyType represents the types of keys associated with encrypting and
+// decrypting session recordings.
+enum RecordingKeyType = {
+  Data = 0;
+  Wrapping = 1;
+}
+
+// Request for RotateSessionRecordingKeys.
+message RotateSessionRecordingKeysRequest {
+  // KeyType defines which key type should be rotated.
+  RecordingKeyType KeyType = 1;
+}
+
+// Response for RotateSessionRecordingKeys
+message RotateSessionRecordingKeysResponse {}
+```
+
+### Session Recording Modes
+
+There are four session recording modes that describe where the recordings are
+captured and how they're shipped to long term storage.
+- `proxy-sync`
+- `proxy`
+- `node-sync`
+- `node`
+Where the recordings are collected is largely unimportant to the encryption
+strategy, but whether or not they are handled async or sync has different
+requirements.
+
+In sync modes the session recording data is written immediately to long term
+storage without intermediate disk writes. This means we can simply instantiate
+an age encryptor at the start of the session and encrypt the recording data as
+it's sent to long term storage.
+
+In async modes the session recording data is written to intermediate `.part`
+files. These files are collected until they're ready for upload and are then
+combined into a single `.tar` file. In order to encrypt individual parts, we
+will build a special `io.Writer` that contains a single instance of the age
+encryptor that can proxy writes across multiple files. This will require
+maintaining a concurrency-safe mapping of in-progress uploads to encrypted
+writers which incurs a bit of added complexity. However it intentionally avoids
+any sort of intermediate key management which seems a worthwhile tradeoff.
```

Comment:
How does this interact with the chunked uploading of parts and the potential for recordings to be terminated while incomplete? AIUI the age format assumes that the file is fully available and must be terminated correctly for the data to be safely decrypted.

Reply:
Edoardo and I talked about this in Slack, but I think it would be possible to to encrypt each chunk separately, allow them to be combined in long term storage, and then split them based on the `age` header at decryption time. We would take a performance hit and a bit of a storage hit, though. I'm still trying to work out if there's a way around that.